### DISCUSSION TITLE
Generic Cloud Backend for Remote Execution

### DISCUSSION THREAD

**Discussion by Smithx10** [opentofu/opentofu#446] [Q&A]
Is there currently a project or server that implements the Cloud Backend for remote execution?

https://opentofu.org/docs/cli/cloud/

I'd love to be able to run my users code remotely in a standardized way.

It would be lovely if there was a gRPC interface that allowed folks to implement the Cloud Backend Interface?

I worked with "[Virtual Kubelet](https://github.com/virtual-kubelet/virtual-kubelet)" and it made it easy for folks to implement their own pod lifecylce logic.

For example the user only had to implement the following interface to deploy pods to their infrastructure.  It would be lovely to have something similar with opentf cloud backends.  

``` go 
type PodLifecycleHandler interface {
    // CreatePod takes a Kubernetes Pod and deploys it within the provider.
    CreatePod(ctx context.Context, pod *corev1.Pod) error

    // UpdatePod takes a Kubernetes Pod and updates it within the provider.
    UpdatePod(ctx context.Context, pod *corev1.Pod) error

    // DeletePod takes a Kubernetes Pod and deletes it from the provider.
    DeletePod(ctx context.Context, pod *corev1.Pod) error

    // GetPod retrieves a pod by name from the provider (can be cached).
    GetPod(ctx context.Context, namespace, name string) (*corev1.Pod, error)

    // GetPodStatus retrieves the status of a pod by name from the provider.
    GetPodStatus(ctx context.Context, namespace, name string) (*corev1.PodStatus, error)

    // GetPods retrieves a list of all pods running on the provider (can be cached).
    GetPods(context.Context) ([]*corev1.Pod, error)
}
```

I've seen abstractions like this,  but they are a bit clunky for the user:
https://github.com/nunchistudio/temporal-land/blob/main/integrations/terraform/documentation/terraform-cdk.md

**Comment by adamlc** [comment]
Maybe [OTF](https://github.com/leg100/otf) is what you're looking for, which is a clone of Terraform Cloud

**Reply by Smithx10** [reply]
Thank you very much for the link,  I had somehow never stumbled upon this project.

**Reply by alfespa17** [reply]
You could also check [terrakube](https://github.com/AzBuilder/terrakube) it support [cloud/remote backend](https://docs.terrakube.io/user-guide/workspaces/cli-driven-workflow)
